### ISSUE TITLE
IoU threshold for training?

### ISSUE THREAD

**Issue by DieBirke** [AlexeyAB/darknet#8448]
Is there a way to modify the loss function so the bounding boxes are more precise?
I know I can calculate map with an IoU threshold (standard 0.5) but can I adjust how the model is trained by inputting such a threshold?

While the model works well enough in detecting my objects, the bounding box isn't quite right sometimes and will cut off small parts of the obejct. Maybe this problem vanishes with more training data, which I am generating atm.
I already looked through my data and the bounding boxes there are as accurate as they can be.

IoU threshold = 0.5 results in a 98% map
IoU threshold = 0.8 drops this down to 45%
Average IoU is 64.58% which is good but I am trying to maximize the precision of the bounding boxes

**Comment by bulatnv** [comment]
Hello @DieBirke 
I'm not sure about loss function or IoU threshold. I can suggest you other ways to improve precision of bounding box:

0) More data will definitely help. 

1) If object in dataset are big enough (bigger than 32x32 pix), add `new_coords=1` flag in the last yolo layer. Read more about it #7645

2) Recalculate anchors or you can even increase the number of anchors. #7856, #6548, #6740

3) Postprocessing: Instead of NMS use something more complex, like WBF (weigted box fusion). Read more about it [here](https://medium.com/analytics-vidhya/weighted-boxes-fusion-86fad2c6be16), [here](https://arxiv.org/pdf/1910.13302.pdf), [here](https://github.com/ZFTurbo/Weighted-Boxes-Fusion)

All the best.

**Comment by DieBirke** [comment]
Hello, thanks for the amount of infos.
Going to try new_coords = 1 for sure as its super easy.

But I have some questions regarding the other methods:

For 2. anchors: I would just add or change the values in the .cfg file for that?

Also I found options in my .cfg file for each yolo layer (where the anchors are incidently) that there is an iou-thresh

```
[yolo]
mask = 6,7,8
anchors = 12, 16, 19, 36, 40, 28, 36, 75, 76, 55, 72, 146, 142, 110, 192, 243, 459, 401
classes=2
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1
scale_x_y = 1.05
iou_thresh=0.213
cls_normalizer=1.0
iou_normalizer=0.07
iou_loss=ciou
nms_kind=greedynms
beta_nms=0.6
max_delta=5
```
So there are options for tweaking the iuo right?
threshold, normalizer and loss. Can you or someone explain these options?

**Comment by bulatnv** [comment]
Most of the parameters in yolo layer are explained in [wiki](https://github.com/AlexeyAB/darknet/wiki/CFG-Parameters-in-the-different-layers)

Unfortunately, I'm not very good at understating loss functions for yolo. This two papers are major papers for bounding box regression training: 
1) [CIoU loss and Cluster NMS](https://arxiv.org/pdf/2005.03572.pdf), 
2) [DIoU loss](https://arxiv.org/pdf/1911.08287.pdf). 
3) This is main repo: [Complete-IoU Loss and Cluster-NMS](https://github.com/Zzh-tju/CIoU)
